File: app/src/main/java/com/softhink/single/registro/view/RegDataFragment.java
package com.softhink.single.registro.view;

import android.app.DatePickerDialog;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RadioGroup;
import com.softhink.single.BaseFragment;
import com.softhink.single.R;
import com.softhink.single.registro.presenter.RegistroContract;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

/**
 * A simple {@link Fragment} subclass.
 */
public class RegDataFragment extends BaseFragment implements
        View.OnClickListener,
        RadioGroup.OnCheckedChangeListener {

    private ImageView btnNext;
    private EditText inputName;
    private EditText txtDay;
    private EditText txtMonth;
    private EditText txtYear;
    private RadioGroup radioGender;
    private Date userBirthday;
    private String gender;
    private RegistroContract.DataContract.CallbackData callback;

    public RegDataFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_registro_uno, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (savedInstanceState == null) {
            inputName = view.findViewById(R.id.inputName);
            txtDay = view.findViewById(R.id.pickerDay);
            txtMonth = view.findViewById(R.id.pickerMonth);
            txtYear = view.findViewById(R.id.pickerYear);
            radioGender = view.findViewById(R.id.radioGender);
            radioGender.setOnCheckedChangeListener(this);
            btnNext = view.findViewById(R.id.btnNextPage);
        }

        btnNext.setOnClickListener(this);
        txtDay.setOnClickListener(this);
        txtMonth.setOnClickListener(this);
        txtYear.setOnClickListener(this);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        try {
            callback = (RegistroContract.DataContract.CallbackData) context;
        } catch (ClassCastException e){
            throw new ClassCastException(getActivity().toString() + " must implements interface");
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.pickerDay:
                showDatePicker();
                break;
            case R.id.pickerMonth:
                showDatePicker();
                break;
            case R.id.pickerYear:
                showDatePicker();
                break;

            case R.id.btnNextPage:
                callback.dataForm(inputName.getText().toString(), userBirthday, gender);
                break;
        }
    }


    @Override
    public void onCheckedChanged(RadioGroup group, int checkedId) {
        switch (checkedId) {
            case R.id.radioFemale:
                gender = "F";
                break;

            case R.id.radioMale:
                gender = "M";
                break;
        }
    }

    private void showDatePicker() {
        Integer[] date = getDateForPicker(txtYear.getText().toString(),
                txtMonth.getText().toString(), txtDay.getText().toString());

        DatePickerDialog datePickerDialog = new DatePickerDialog(getActivity(), new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                userBirthday = getUserBirthDay(year, month, dayOfMonth);

                txtDay.setText((dayOfMonth < 10)?getString(R.string.date_mask, dayOfMonth):
                        String.valueOf(dayOfMonth));
                txtMonth.setText((month < 9)?getString(R.string.date_mask, month+1):String.valueOf(month));
                txtYear.setText(String.valueOf(year));
            }
        }, date[0], date[1], date[2]);

        datePickerDialog.show();
    }

    private Integer[] getDateForPicker(String s1, String s2, String s3) {
        if (s1.isEmpty() && s2.isEmpty() && s3.isEmpty()){
            return new Integer[]{Calendar.getInstance().get(Calendar.YEAR),
                    Calendar.getInstance().get(Calendar.MONTH),
                    Calendar.getInstance().get(Calendar.DAY_OF_MONTH)};
        } else {
            return new Integer[]{Integer.valueOf(s1), Integer.valueOf(s2) -1, Integer.valueOf(s3)};
        }
    }

    private Date getUserBirthDay(int year, int month, int dayOfMonth) {
        try {
            return new SimpleDateFormat("yyyy-MM-dd", new Locale("es"))
                    .parse(""+year+"-"+(month+1)+"-"+dayOfMonth);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: app/src/main/java/com/softhink/single/registro/view/RegDataFragment.kt
package com.softhink.single.registro.view

import android.app.AlertDialog
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.DatePicker
import android.widget.RadioGroup
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.softhink.single.base.BaseFragment
import com.softhink.single.R
import com.softhink.single.registro.SignUpViewModel
import com.softhink.single.registro.Status.*
import kotlinx.android.synthetic.main.arrow_next.*
import kotlinx.android.synthetic.main.fragment_registro_uno.*
import java.lang.Exception
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * A simple [Fragment] subclass.
 */
class RegDataFragment : BaseFragment(), View.OnClickListener, RadioGroup.OnCheckedChangeListener {

    private var userBirthday: Date? = null
    private var gender: String? = null
    private lateinit var model: SignUpViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        model = activity?.run {
            ViewModelProviders.of(this).get(SignUpViewModel::class.java)
        }?:throw Exception("Invalid Activity")
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_registro_uno, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (savedInstanceState == null) {
            btnNextPage.setOnClickListener(this)
            pickerDay.setOnClickListener(this)
            pickerMonth.setOnClickListener(this)
            pickerYear.setOnClickListener(this)
            radioGender.setOnCheckedChangeListener(this)
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.pickerDay -> showDatePicker()
            R.id.pickerMonth -> showDatePicker()
            R.id.pickerYear -> showDatePicker()

            R.id.btnNextPage -> model.validateForm(inputName.text?.trim().toString(),
                    userBirthday, gender).observe(this,
                    Observer {
                        when(it.status){
                            ERROR -> showMessageDialog(getString(it.data))
                            SUCCESS -> toAccountFragment()
                            FAILED -> TODO()
                        }
                    })
        }
    }


    override fun onCheckedChanged(group: RadioGroup, checkedId: Int) {
        when (checkedId) {
            R.id.radioFemale -> gender = "F"

            R.id.radioMale -> gender = "M"
        }
    }

    private fun showDatePicker() {
        val inflater = activity?.layoutInflater

        val view = inflater?.inflate(R.layout.picker_layout, null)
        val datePicker = view?.findViewById<DatePicker>(R.id.pickerDate)

        val builder = AlertDialog.Builder(activity)
        builder.setView(view)
        builder.setPositiveButton("ACEPTAR") {
            dialogInterface, i ->
            userBirthday = getUserBirthDay(datePicker?.year!!,
                    datePicker.month, datePicker.dayOfMonth)

            pickerDay.setText(datePicker.dayOfMonth.toString())
            pickerMonth.setText((datePicker.month + 1).toString())
            pickerYear.setText(datePicker.year.toString())
        }
        builder.setNegativeButton("CANCELAR", null)
        builder.create().show()
    }

    private fun getUserBirthDay(year: Int, month: Int, dayOfMonth: Int): Date? {
        try {
            return SimpleDateFormat("yyyy-MM-dd", Locale("es"))
                    .parse("" + year + "-" + (month + 1) + "-" + dayOfMonth)
        } catch (e: ParseException) {
            e.printStackTrace()
        }

        return null
    }

    private fun toAccountFragment(){
        fragmentManager?.beginTransaction()
                ?.add(R.id.registroContainer,
                        RegAccountFragment(), RegAccountFragment::class.java.simpleName)
                ?.addToBackStack(RegAccountFragment::class.java.simpleName)
                ?.commit()
    }
}